### PR TITLE
CI: Move some Cirrus CI jobs to GitHub Actions

### PR DESCRIPTION
**Pull request by kinke**

### CODE REVIEW

**Comment by kinke** [comment]
Not sure if I should better target `stable` instead, let me know.

Otherwise simpler than anticipated and working as intended. Migrating the coverage jobs needs a bit more work, incl. setting a secret env var, so better done separately in a follow-up PR. FreeBSD isn't natively supported by the hosted GitHub Actions runners.

The branch settings in the GitHub UI need to be adapted after the merge, accounting for the slightly different macOS job names (as 'required checks').

**Comment by ibuclaw** [comment]
> Not sure if I should better target `stable` instead, let me know.
> 
> Otherwise simpler than anticipated and working as intended. Migrating the coverage jobs needs a bit more work, incl. setting a secret env var, so better done separately in a follow-up PR. FreeBSD isn't natively supported by the hosted GitHub Actions runners.
> 
> The branch settings in the GitHub UI need to be adapted after the merge, accounting for the slightly different macOS job names (as 'required checks').

Master, as I'm waiting on merging master->stable.

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -0,0 +1,93 @@
+# https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions
+
+name: Main
+on:
+  - pull_request # without merge conflicts
+  - push # branch or tag
```
Comment (ljmf00) [inline .github/workflows/main.yml:6]: Should probably specify the target branch

**Comment by ibuclaw** [comment]
> FreeBSD isn't natively supported by the hosted GitHub Actions runners.
> 

This Action plugin/addon? 

https://github.com/vmactions

**Comment by kinke** [comment]
> This Action plugin/addon?

I guess one of these VM workarounds could be a (not-so-nice) option, requiring some boilerplate YAML (AFAIK, GitHub has its own parser without YAML anchors support, so e.g. can't mixin some 'common steps' defined in the same file into a VM step wrapper as shown in https://github.com/vmactions/freebsd-vm).

But that's outside the scope of this PR.

**Comment by ibuclaw** [comment]
> > This Action plugin/addon?
> 
> **Edit**: Oh, based on VirtualBox on Mac - the Mac hosts performance is pretty low already. qemu on Linux would probably be better.
> **Edit2**: And there's an extra max concurrency limit of 5 for Mac runners, vs. 20 concurrent runners in general: https://docs.github.com/en/actions/learn-github-actions/usage-limits-billing-and-administration#usage-limits

I assume github allows self-hosted agents? Does it support bsd?

**Comment by kinke** [comment]
> I assume github allows self-hosted agents?

Yes.

> Does it support bsd?

Not officially - Linux, Windows and macOS only: https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners/about-self-hosted-runners#supported-architectures-and-operating-systems-for-self-hosted-runners

But there's e.g. https://github.com/ChristopherHX/github-act-runner, which should run on FreeBSD.

IIUC correctly, Jacob's https://github.com/marketplace/actions/cross-platform-action#runners would support Linux hosts and qemu VMs for FreeBSD.

**Comment by ibuclaw** [comment]
> But there's e.g. https://github.com/ChristopherHX/github-act-runner, which should run on FreeBSD.
> 
> IIUC correctly, Jacob's https://github.com/marketplace/actions/cross-platform-action#runners would support Linux hosts and qemu VMs for FreeBSD.

Ok, so there's a couple options. I could spin up a fbsd VM to test the first project, though the second seems more future proof.

**Comment by ibuclaw** [comment]
> Ok, so there's a couple options. I could spin up a fbsd VM to test the first project, though the second seems more future proof.

Until that is until we run out of more compute resources. 🙄

**Review by ibuclaw — APPROVED** [review]

**Comment by kinke** [comment]
> The branch settings in the GitHub UI need to be adapted after the merge, accounting for the slightly different macOS job names (as 'required checks').

Done.

**Pull request merged**